src/io/simulator/simulator.hpp

```diff
@@ -0,0 +1,44 @@
+// Copyright 2022 Memgraph Ltd.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt; by using this file, you agree to be bound by the terms of the Business Source
+// License, and you may not use this file except in compliance with the Business Source License.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+#pragma once
+
+#include <memory>
+
+#include "io/address.hpp"
+#include "io/simulator/simulator_config.hpp"
+#include "io/simulator/simulator_handle.hpp"
+#include "io/simulator/simulator_transport.hpp"
+
+namespace memgraph::io::simulator {
+class Simulator {
+  std::mt19937 rng_{};
+  std::shared_ptr<SimulatorHandle> simulator_handle_;
+
+ public:
+  explicit Simulator(SimulatorConfig config)
+      : rng_(std::mt19937{config.rng_seed}), simulator_handle_{std::make_shared<SimulatorHandle>(config)} {}
+
+  void ShutDown() { simulator_handle_->ShutDown(); }
+
+  Io<SimulatorTransport> Register(Address address) {
```

Comment:
Why pass by value?

Reply:
because it's just a few bytes on the stack, and can be correctly memcpy'd without loss of correctness. My assumption is that anything that does not have associated allocations or other resources, and memcpy-friendly, is better to pass by value and let the compiler narrow down the specifics that it should push into the stackframe.

Reply:
We have a thumb of rule is to pass objects by reference if they are bigger than `2 * sizeof(pointer)`. If you specify to pass an object by value, then the compiler will always copy, it cannot optimize out the copying. However this case is fine, as
 the [size of Address](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:10,endLineNumber:26,positionColumn:10,positionLineNumber:26,selectionStartColumn:10,selectionStartLineNumber:26,startColumn:10,startLineNumber:26),source:'%23include+%3Ciostream%3E%0A%23include+%3Cmemory%3E%0A%23include+%3Cstring%3E%0A%23include+%3Cvariant%3E%0A%0A%0A%23include+%3Cboost/asio/ip/tcp.hpp%3E%0A%23include+%3Cboost/uuid/uuid.hpp%3E%0A%23include+%3Cboost/uuid/uuid_io.hpp%3E%0A%0Astruct+Address+%7B%0A++//+It!'s+important+for+all+participants+to+have+a%0A++//+unique+identifier+-+IP+and+port+alone+are+not%0A++//+enough,+and+may+change+over+the+lifecycle+of%0A++//+the+nodes.+Particularly+storage+nodes+may+change%0A++//+their+IP+addresses+over+time,+and+the+system%0A++//+should+gracefully+update+its+information%0A++//+about+them.%0A++boost::uuids::uuid+unique_id%3B%0A++boost::asio::ip::address+last_known_ip%3B%0A++uint16_t+last_known_port%3B%0A%7D%3B%0A%0Aint+main()+%7B%0Astd::cout+%3C%3C+sizeof(Address)+%3C%3C+std::endl%3B%0Areturn+0%3B%0A%7D'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:50,l:'4',m:50,n:'0',o:'',s:0,t:'0'),(g:!((g:!((h:compiler,i:(compiler:clang1301,filters:(b:'0',binary:'0',commentOnly:'0',demangle:'1',directives:'0',execute:'0',intel:'1',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!((name:boost,ver:'179')),options:'--std%3Dc%2B%2B20+',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1,tree:'1'),l:'5',n:'0',o:'x86-64+clang+13.0.1+(C%2B%2B,+Editor+%231,+Compiler+%231)',t:'0')),header:(),l:'4',m:50,n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'x86-64+gcc+12.1',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+clang+13.0.1+(Compiler+%231)',t:'0')),header:(),k:50,l:'4',m:50,n:'0',o:'',s:0,t:'0')),k:50,l:'3',n:'0',o:'',t:'0')),l:'2',m:100,n:'0',o:'',t:'0')),version:4) is only 48 bytes.
 
And I lied about not being able to optimize out: for function that in can inline (mostly for functions defined in the same translation unit) the compiler can actually super smart, but not in general.